Core IR to Procedural IR

sthaeron opened the pull request:
Adds functionality to translate IRFunctions from ForSyDe IR into Procedural IR.
- Adds a module which translates Core IR function expressions into Procedural IR
- Adds various SDF examples used for testing
- Makes C the default output of the compiler

sthaeron committed:
feat: adds simple SDF examples to start Core IR to Procedural IR translation

sthaeron committed:
fix: pretty printer adds unique names to binders

sthaeron committed:
feat: starts ForSyDe IR to Procedural IR translation, by adding dynamic function declaration

sthaeron committed:
feat: adds translation of function core expressions to procedural IR

sthaeron committed:
feat: adds new SDF examples to test new dynamic core function expression translation

sthaeron committed:
feat: updates translation of function core expressions to use folds

sthaeron committed:
feat: makes C the default output of compiler

SamuelMiksits commented:
```C
for (int i = 0; i < 2; i++) {
    status = scanf("%d", &input_s_in_1[i]);
}
if (status < 1) break;
for (int i = 0; i < 2; i++) {
    write_token(s_in_1, input_s_in_1[i]);
}
```
oopsie ("""someone""" changed if so that it now puts break on the same line)

SamuelMiksits commented:
> ```c
> for (int i = 0; i < 2; i++) {
>     status = scanf("%d", &input_s_in_1[i]);
> }
> if (status < 1) break;
> for (int i = 0; i < 2; i++) {
>     write_token(s_in_1, input_s_in_1[i]);
> }
> ```
> 
> oopsie ("""someone""" changed if so that it now puts break on the same line)

so basically this
```diff
diff --git a/src/ForSyDeIRToProceduralIR.hs b/src/ForSyDeIRToProceduralIR.hs
index e3b0399..df4c4d0 100644
--- a/src/ForSyDeIRToProceduralIR.hs
+++ b/src/ForSyDeIRToProceduralIR.hs
@@ -203,7 +203,7 @@ translateIRConstructor initialContext constructor = case constructor of
                     (EBinOp Less (EVar "i") (EInt (bufferSize)))
                     (SExpr (EUnOp Increment (EVar "i")))
                     (SScope [SVarAssign "status" (ECall "scanf" [EString "%d", EReference (EArrayAccess (EVar ("input_" ++ show signalId)) (EVar "i"))])])
-                breakIfStmt = SIf (EBinOp Less (EVar "status") (EInt 1)) (SBreak) Nothing
+                breakIfStmt = SIf (EBinOp Less (EVar "status") (EInt 1)) (SScope [SBreak]) Nothing
                 writeForStmt =
                   SFor
                     (SVarDef TInt "i" (EInt 0))
```

SamuelMiksits commented on examples/model/SDF_example_011.hs:
```diff
@@ -0,0 +1,17 @@
+
+import ForSyDe.Shallow
+
+-- Net list
```
I would say we are about 50/50 having
```hs
-- Netlist
```
and
```hs
-- Net list
```
Should we do something about this?

sthaeron replied:
Based on the lecture notes it should be netlist

SamuelMiksits commented on src/ForSyDeIRToProceduralIR.hs:
```diff
@@ -56,7 +57,7 @@ translateIRSystemToProgram dflags scheduleList bufferList delayBufferList lookup
       context1 = foldl' translateIRConstructor initialContext constructors
       context2 = foldl' translateBuffer context1 bufferList
-      context3 = foldl' translateIRFunctionToGlobals context2 functionList
+      context3 = foldl' (translateIRFunctionToGlobals constructors) context2 functionList
```
Might be fine here since it has minimal impact, but I moved variables like `runs` and `inputType` into translation context because otherwise you would need to perform massive partial applications for the fold, i.e:

```hs
context3 = foldl' (translateIRFunctionToGlobals constructors runs inputType) context2 functionList
```

and this defeats the purpose of having translation context in the first place

SamuelMiksits commented on src/CoreIRToProceduralIR.hs:
```diff
@@ -0,0 +1,277 @@
+    let context1 = context {functionInputs = [(IRVar b)]}
+     in translateFunctionExpr context1 inputIndex inputCounter e
+  Let _ e -> translateFunctionExpr context inputIndex inputCounter e
+  Case (Var i) _ _ty alts -> case elemIndex (IRVar i) (functionInputs context) of
```
It is clear from the other comments when the cases are invoked, but what is this case?

SamuelMiksits reviewed (commented):
I think it looks fine in general. Could just do a few more comments in some places.

SamuelMiksits commented on src/CoreIRToProceduralIR.hs:
```diff
@@ -0,0 +1,277 @@
+findActorFromFunctionId :: IRId -> [IRConstructor] -> Maybe IRConstructor
+findActorFromFunctionId targetFunctionId constructors =
+  let checkIRConstructor :: IRConstructor -> Maybe IRConstructor
+      checkIRConstructor actor@(IRActor _ _ functionId _)
```
I think the usage of the "@" here and in FPlus, etc is pretty esoteric and non-obvious and warrants some explanation

klarasm commented on src/CoreIR.hs:
```diff
@@ -29,13 +35,13 @@ prettyCoreProgram dflags = intercalate "\n" . map (prettyCoreBind dflags)
   Lit l -> printf "Lit(%s)" (showPpr dflags l)
   App e a -> printf "App(%s *\n%s)" (prettyCoreExpr dflags e) (prettyCoreExpr dflags a)
-  Lam b e -> printf "Lam(%s ->\n%s)" (varToString b) (prettyCoreExpr dflags e)
+  Lam b e -> printf "Lam(%s ->\n%s)" (showPpr dflags b) (prettyCoreExpr dflags e)
```
Is there a reason we go back to showPpr here? Is it he unique stuff (you add a varUnique above)?

sthaeron replied:
We were having problems with the Core IR pretty printer, so we decided to revert to what we originally had. Especially since all the other cases are using `showPpr` anyways.

klarasm commented on src/CoreIRToProceduralIR.hs:
```diff
@@ -0,0 +1,277 @@
+   in case function of
+        IRFunction functionId Nothing ->
+          let actor = case (findActorFromFunctionId functionId constructors) of
+                Nothing -> error ("translateIRFunction - no actor associated to " ++ show functionId)
```
Is there a reason we error if there's no actor associated here and below? As we're making them static it would just not be in the binary. Would it be possible to just process it normally and don't crash the compiler on it?

klarasm replied:
Right, you get the parameter list from the actor type? In that case it might be a little more work and I would ve fine keeping it this way.

sthaeron replied:
I think we could potentially revisit this if we decide to redo how we handle errors. For now, this adds the programmer restriction that all functions must have an associated actor which calls them.

klarasm commented on src/CoreIRToProceduralIR.hs:
```diff
@@ -0,0 +1,277 @@
+           in (functionDeclarationGlobal, Nothing)
+        IRFunction functionId (Just functionExpr) ->
+          let actor = case (findActorFromFunctionId functionId constructors) of
+                Nothing -> error ("translateIRFunction - no actor associated to " ++ show functionId)
```
Here too.

klarasm commented on src/ForSyDeIR.hs:
```diff
@@ -42,6 +43,7 @@ data IRId
   | Empty
 
 instance Show IRId where
+  show :: IRId -> String
```
This seems a little unrelated, any reason for it?

sthaeron replied:
It's just adding a type to the function instance for consistency with the rest of the code base.

klarasm reviewed (changes requested):
Looks good in general, but do we really need to crash if we can't associate a function to an actor?

SamuelMiksits commented on examples/model/SDF_example_020.hs:
I already added the example from the labs as `examples/model/SDF_example_002.hs`, that file is exactly the same

sthaeron replied:
Missed that, I'll drop the commit then.

sthaeron committed:
refactor: Updates netlist to be a single word in examples, comments, and docs

sthaeron committed:
refactor: adds comments and cleans up Core IR to Procedural IR translation

sthaeron commented:
@SamuelMiksits and @klarasm, I've added comments to the Core IR to Procedural IR translation solution and cleaned up some of the functions in the hopes that they are more readable.

SamuelMiksits reviewed (approved):
Yeah I think I am happy enough with this now

klarasm reviewed (approved):
Looks good!
